rfd/0138-postgres-backend.md

```diff
@@ -0,0 +1,121 @@
+
+With minor modifications, this schema can support range based partitioning by `event_time`, creating partitions by hand or using `pg_partman`. The initial implementation will not use partitioning, but setting up the table layout for it will allow for a much easier migration if and when we end up supporting it.
+
+### Requirements and recommendations
+
+Any supported version of Postgres (11 to 15) should work, with `wal2json` 2.1 or later, although for better results, Postgres 13 or higher is recommended (as logical decoding had some internal improvements).
+
+Because the db user that Teleport connects as needs `REPLICATION` permissions, which effectively grants read access to the entire database cluster, the backend should not be used in a database cluster that also contains data that the Teleport auth server should not have access to (it's fine to use a single cluster for the backend and the audit log storage, for instance).
+
+Because of how temporary replication slots (and prepared statements) are tied to individual connections, the use of external connection poolers (such as `pgbouncer`) is not supported.
+
+For self-managed Postgres setups we recommend the use of [certificate authentication](https://www.postgresql.org/docs/current/auth-cert.html); in Azure we support and recommend the use of Azure AD authentication over the use of hardcoded passwords.
+
+Teleport will attempt to create the databases specified in the configuration for itself, and set them up with the correct schema (as well as apply any future migrations, if needed). It's possible to create the databases manually and not give Teleport permissions to create databases, and technically it's even possible to manually set up the database schemas (to avoid granting Teleport the ability to create tables, for instance), but we recommend giving Teleport its own database for the backend - and since Teleport will need read-write access to all the data in it anyway, we also recommend letting Teleport manage the schema automatically.
+
+It's possible to only grant `INSERT` and `SELECT` privileges to the Teleport db user on the table that stores audit events, but not UPDATE or DELETE privileges; such a setup couldn't be automatically managed by Teleport, however, and thus would require detailed documentation on how to set up the schema, and some care.
+
+For reliability, it's recommended to run a standby server with failover and synchronous replication (by setting `synchronous_standby_names`). The details on how to do this vary depending on the way Postgres is deployed; for instance, on Azure it's sufficient to enable ["High availability" mode](https://learn.microsoft.com/en-us/azure/postgresql/flexible-server/concepts-high-availability).
+
+### UX
```

Comment:
Open UX questions:
- which tunables should we expose? (delete batch size, expiry batch size, expiry interval, disable expiry loop, event poll interval, event batch size)
- the conn_string (as well as the rest of the pgx/pgxpool machinery) can use a password, either embedded or by reading it from the `~/.pgpass` file. Should we document that? Should we detect it and deliberately refuse to proceed with a password? What about TLS options such as `sslmode=verify-full`?
- should we require an option for the "change feed mode" even if we currently only support `wal2json` with the SQL interface? Requiring an option to be set might be better in the long term compared to having a default that we might deprecate in the future.